pandas/core/indexing.py

```diff
@@ -971,65 +951,48 @@ def _reindex(keys, level=None):
             else:
                 level = None
 
-            keyarr_is_unique = Index(keyarr).is_unique
+            # existing labels are unique and indexer are unique
+            if labels.is_unique and Index(keyarr).is_unique:
+
+                try:
+                    result = self.obj.reindex_axis(keyarr, axis=axis, level=level)
+
+                    # this is an error as we are trying to find
+                    # keys in a multi-index that don't exist
+                    if isinstance(labels, MultiIndex) and level is not None:
+                        if hasattr(result,'ndim') and not np.prod(result.shape) and len(keyarr):
+                            raise KeyError("cannot index a multi-index axis with these keys")
+
+                    return result
 
-            # existing labels are unique and indexer is unique
-            if labels.is_unique and keyarr_is_unique:
-                return _reindex(keyarr, level=level)
+                except AttributeError:
```

Comment:
you have a lot in the `try` block here. Can you use an `else` clause to limit the amount of code the attribute error checks?